File: No_2439/Minimize Maximum of Array.cpp
class Solution {
public:
    bool checkSatisfied(vector<int>& nums, int target) {
        long long carry = 0;
        for (int i = nums.size() - 1; i >= 1; i--) {
            if (nums[i] + carry <= target) {
                carry = 0;
                continue;
            }
            carry = (nums[i] + carry) - target;
        }
        
        return nums[0] + carry <= target ? true : false;
    }
    
    int minimizeArrayValue(vector<int>& nums) {
        int low = 0, high = INT_MAX;
        while (low < high) {
            int mid = low + (high - low) / 2;
            if (checkSatisfied(nums, mid))
                high = mid;
            else
                low = mid + 1;
        }
        
        return high;
    }
};

File: No_2439/Minimize Maximum of Array.py
class Solution:
    def minimizeArrayValue(self, nums: List[int]) -> int:
        def checkSatisfied(nums, target):
            carry = 0
            for num in nums[:0:-1]:
                if carry + num <= target:
                    carry = 0
                else:
                    carry = num + carry - target
            
            return True if nums[0] + carry <= target else False
        
        low = 0
        high = int(1e9)
        
        while low < high:
            mid = low + (high - low) // 2
            if checkSatisfied(nums, mid):
                high = mid
            else:
                low = mid + 1
                
        return high